File: go/raindrops/raindrops.go
// Package raindrops implements the pling-plong child's game
package raindrops

import (
	"strconv"
)

// Convert a n into its pling-plong equivalent depending on its divisors
func Convert(n int) string {
	var equivalent string
	if n%3 == 0 {
		equivalent += "Pling"
	}
	if n%5 == 0 {
		equivalent += "Plang"
	}
	if n%7 == 0 {
		equivalent += "Plong"
	}
	if equivalent == "" {
		equivalent = strconv.Itoa(n)
	}
	return equivalent
}

File: haskell/raindrops/src/Raindrops.hs
module Raindrops (convert) where

convert :: Int -> String
convert n = case plingPlangWords n of
  [] -> show n
  ws -> concat ws
  
plingPlangWords :: Int -> [String]
plingPlangWords n = [w | (d, w) <- plingPlangRules,
                         n `divisibleBy` d ]

plingPlangRules :: [(Int, String)]
plingPlangRules = [(3, "Pling"), (5, "Plang"), (7, "Plong")]

divisibleBy :: Int -> Int -> Bool
divisibleBy x y = x `rem` y == 0
